File: leetcode/java/0019. Remove Nth Node From End of List.java
package leetcode.java;

// Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

// Time complexity: O(N)
// Space complexity: O(1)

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int numElements = 0;
        ListNode temp = head;
        while (temp != null) {
            temp = temp.next;
            numElements += 1;
        }

        int toLoop = numElements - n - 1;
        if (toLoop < 0) {
            return head.next;
        }

        temp = head;
        while (toLoop > 0) {
            temp = temp.next;
            toLoop -= 1;
        }
        temp.next = temp.next.next;
        return head;
    }
}

File: leetcode/python/0019. Remove Nth Node From End of List.py
# Time complexity: O(N)
# Space complexity: O(1)

class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution(object):
    def removeNthFromEnd(self, head, n):
        length = 0
        temp = head
        
        while temp != None:
            temp = temp.next
            length += 1

        if length == 1:
            return None
        
        toLoop = length - n - 1
        if toLoop == -1:
            return head.next
        
        temp = head
        for _ in range(toLoop):
            temp = temp.next

        temp.next = temp.next.next
        return head